[clang][bytecode] Handle negative array sizes in constexpr `new` instead of asserting

snarang181 opened the pull request:
This patch fixes a crash in the constexpr interpreter when evaluating `new T[n]` with a negative element count.

Fixes https://github.com/llvm/llvm-project/issues/152904

snarang181 committed:
Enable nullptr handle with negative elemsize in
a dynamic allocation

frederick-vs-ja commented on clang/test/SemaCXX/new-neg-size.cpp:
```diff
@@ -0,0 +1,15 @@
+// In C++20, constexpr dynamic allocation is permitted *only* if valid.
+// A negative element count must be diagnosed (and must not crash).
+
+constexpr void f_bad_neg() {
+  int a = -1;
+  (void) new int[a]; // triggers negative-size path in the interpreter
+}
```
It seems to me that we should also cover `new (std::nothrow) int[a]`, in which case a null pointer should be returned.

snarang181 replied:
Added

snarang181 committed:
Add test case

snarang181 committed:
Revert "Add test case"

This reverts commit 6d28107ebb963803b3bf6c4ae09734ec1ec6b526.

tbaederr commented on clang/test/SemaCXX/new-neg-size.cpp:
```diff
@@ -0,0 +1,15 @@
+// RUN: not %clang_cc1 -std=c++20 -fsyntax-only %s 2>&1 \
+// RUN:   | FileCheck %s --implicit-check-not='Assertion `NumElements.isPositive()` failed'
+
```
You don't use `FileCheck` for these kinds of tests, use `-verify`.

tbaederr replied:
Oh and the test should go into `test/AST/ByteCode/`

snarang181 committed:
Add nothrow test for negative size allocation in constant expressions
Move test directory to AST/
Add Diagnostic Note for negative size allocation in constant expressions

snarang181 committed:
Move test cases into new-delete

snarang181 committed:
Shift expected output to line it is emitted on

snarang181 commented on clang/test/AST/ByteCode/new-delete.cpp:
```diff
@@ -1070,14 +1070,13 @@ namespace BaseCompare {
   int x = -1;
-  int *p = new int[x]; // expected-note {{cannot allocate array with negative size in a constant expression}}
+  int *p = new int[x]; // expected-note {{cannot allocate array with negative size in a constant expression}} \
+                       // ref-note {{cannot allocate array; evaluated array bound -1 is negative}}
```
Without the byte code interpreter, the diagnostic is not emitted. 
` note: cannot allocate array; evaluated array bound -1 is negative`
However, with the `-fexperimental-new-constant-interpreter` flag,  we hit the diagnostic. 
`note: cannot allocate array with negative size in a constant expression`

snarang181 replied:
I changed the Bytecode interpreter to reuse the same diagnostic now; I had created a new one earlier. There's coherent output between both the interpreters now.

snarang181 committed:
Change new interp error to reuse legacy error

tbaederr commented on clang/lib/AST/ByteCode/Interp.h:
```diff
@@ -3490,7 +3490,15 @@ inline bool AllocN(InterpState &S, CodePtr OpPC, PrimType T, const Expr *Source,
     return true;
   }
-  assert(NumElements.isPositive());
+  if (!NumElements.isPositive()) {
```
```suggestion
  if (NumElements.isNegative()) {
```

tbaederr commented on clang/lib/AST/ByteCode/Interp.h:
```diff
@@ -3490,7 +3490,15 @@ inline bool AllocN(InterpState &S, CodePtr OpPC, PrimType T, const Expr *Source,
-  assert(NumElements.isPositive());
+  if (!NumElements.isPositive()) {
+    if (!IsNoThrow) {
+      S.FFDiag(Source, diag::note_constexpr_new_negative)
```
```suggestion
      S.FFDiag(S.Current->getSource(OpPC), diag::note_constexpr_new_negative)
```

tbaederr commented on clang/test/AST/ByteCode/new-delete.cpp:
```diff
@@ -1069,6 +1069,30 @@ namespace BaseCompare {
 }
 
+
+namespace NegativeArraySize { 
+
+constexpr void f() { // both-error {{constexpr function never produces a constant expression}}
+  int x = -1;
+  int *p = new int[x]; //both-note {{cannot allocate array; evaluated array bound -1 is negative}} 
+}
+
+} // namespace NegativeArraySize
+
```
```suggestion
namespace NegativeArraySize { 
  constexpr void f() { // both-error {{constexpr function never produces a constant expression}}
    int x = -1;
    int *p = new int[x]; //both-note {{cannot allocate array; evaluated array bound -1 is negative}} 
  }
} // namespace NegativeArraySize

```

snarang181 committed:
Address feedback
